File: core/src/com/unciv/models/gamebasics/TileImprovement.java
package com.unciv.models.gamebasics;

import com.unciv.logic.civilization.CivilizationInfo;
import com.unciv.models.stats.FullStats;
import com.unciv.models.stats.NamedStats;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;

public class TileImprovement extends NamedStats implements ICivilopedia {
    public Collection<String> terrainsCanBeBuiltOn = new ArrayList<String>();
    public String techRequired;

    public String improvingTech;
    public FullStats improvingTechStats;

    private int turnsToBuild; // This is the base cost.
    public int getTurnsToBuild(){
        float realTurnsToBuild = turnsToBuild;
        if(CivilizationInfo.current().getBuildingUniques().contains("WorkerConstruction"))
            realTurnsToBuild*=0.75;
        if(CivilizationInfo.current().policies.isAdopted("Citizenship"))
            realTurnsToBuild*=0.75;
        return Math.round(realTurnsToBuild);
    }

    @Override
    public String getDescription() {
        StringBuilder stringBuilder = new StringBuilder();
        if(!new FullStats(this).toString().isEmpty()) stringBuilder.append(new FullStats(this)+"\r\n");
        if(!terrainsCanBeBuiltOn.isEmpty()) stringBuilder.append("Can be built on " + StringUtils.join(", ", terrainsCanBeBuiltOn));

        HashMap<String,ArrayList<String>> statsToResourceNames = new HashMap<String, ArrayList<String>>();
        for(TileResource tr : GameBasics.TileResources.values()){
            if(!tr.improvement.equals(name)) continue;
            String statsString = tr.improvementStats.toString();
            if(!statsToResourceNames.containsKey(statsString))
                statsToResourceNames.put(statsString,new ArrayList<String>());
            statsToResourceNames.get(statsString).add(tr.name);
        }
        for(String statsString : statsToResourceNames.keySet()){
            stringBuilder.append("\r\n"+statsString+" for "+ StringUtils.join(", ",statsToResourceNames.get(statsString)));
        }

        if(techRequired !=null) stringBuilder.append("\r\nTech required: "+ techRequired);

        return stringBuilder.toString();
    }
}

File: core/src/com/unciv/models/gamebasics/TileImprovement.kt
package com.unciv.models.gamebasics

import com.unciv.logic.civilization.CivilizationInfo
import com.unciv.models.stats.FullStats
import com.unciv.models.stats.NamedStats

import java.util.ArrayList
import java.util.HashMap

class TileImprovement : NamedStats(), ICivilopedia {

    @JvmField var terrainsCanBeBuiltOn: Collection<String> = ArrayList()
    @JvmField var techRequired: String? = null

    @JvmField var improvingTech: String? = null
    @JvmField var improvingTechStats: FullStats? = null

    private val turnsToBuild: Int = 0 // This is the base cost.
    fun getTurnsToBuild(civInfo: CivilizationInfo): Int {
        var realTurnsToBuild = turnsToBuild.toFloat()
        if (civInfo.buildingUniques.contains("WorkerConstruction"))
            realTurnsToBuild *= 0.75f
        if (civInfo.policies.isAdopted("Citizenship"))
            realTurnsToBuild *= 0.75f
        return Math.round(realTurnsToBuild)
    }

    override val description: String
        get() {
            val stringBuilder = StringBuilder()
            if (!this.clone().toString().isEmpty()) stringBuilder.appendln(this.clone().toString())
            if (!terrainsCanBeBuiltOn.isEmpty()) stringBuilder.appendln("Can be built on " + terrainsCanBeBuiltOn.joinToString(", "))

            val statsToResourceNames = HashMap<String, ArrayList<String>>()
            for (tr: TileResource in GameBasics.TileResources.values.filter { it.improvement == name }) {
                val statsString = tr.improvementStats.toString()
                if (!statsToResourceNames.containsKey(statsString))
                    statsToResourceNames[statsString] = ArrayList()
                statsToResourceNames[statsString]!!.add(tr.name)
            }
            statsToResourceNames.forEach {
                stringBuilder.appendln(it.key + " for " + it.value.joinToString(", "))
            }

            if (techRequired != null) stringBuilder.appendln("Tech required: " + techRequired)

            return stringBuilder.toString()
        }
}